### PR TITLE
node,rpc: JWT auth dialing + auth endpoints getters + auth testing

### PR DESCRIPTION
**Pull request by protolambda**
This PR:
- Implements testing of the RPC authentication conditions as specified by the Engine API authentication spec
- Adds getters to the node to retrieve the WS and HTTP auth endpoints (useful for testing when the port is assigned dynamically, mirrors the existing getters for non-auth endpoints)
- Updates an old comment that still referred to JWT secrets being passed by direct cli arg instead of cli arg to file.
- Implements `rpc.DialWithAuth`, `rpc.DialHTTPWithAuth`, `rpc.DialWebsocketWithAuth` that take a `rpc.HeaderAuthProvider` and implemented with `rpc.JWTAuthProvider` (along with simple test-only implementations). Thanks to @trianglesphere for implementing this (Optimism uses the Engine API and authentication too, and we want to contribute things back upstream :) )
  - I have thought about using an "option" pattern too (`rpc.DialContext(ctx, endpoint, options... Option)` where one `Option` can add authentication. But the authentication is specific to http and websockets, so the option pattern would be weird with IPC etc. dials. And adding optional args to the existing `DialContext` might break some users that rely on the function type (regular callers wouldn't be affected). So I didn't use this pattern.
- Tests the authentication:
  - That the engine and eth modules are available
  - That basic auth works
  - That other secrets don't work
  - That the "none" algorithm trick doesn't work
  - That the issued-at time is verified with the spec 60 second margin
  - That repeated http calls are also authenticated
  - With all of the above, that the dial-with-auth of the rpc client works

### CODE REVIEW

File: node/node_auth_test.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+		HTTPHost:  "127.0.0.1",
+		HTTPPort:  0,
+		WSHost:    "127.0.0.1",
+		WSPort:    0,
+		AuthAddr:  "127.0.0.1",
+		AuthPort:  0,
+		JWTSecret: jwtPath,
+
+		WSModules:   []string{"eth", "engine"},
+		HTTPModules: []string{"eth", "engine"},
```
Comment (protolambda) [inline node/node_auth_test.go:105]: I believe these are actually ignored by the `Node.startRPC.initAuth` func. Instead it uses `DefaultAuthModules`, and doesn't make them configurable yet. Should we make the exposed auth modules configurable?

File: node/node_auth_test.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+	"fmt"
+	"net/http"
+	"os"
+	"path"
+	"testing"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/rpc"
+
```
Comment (holiman) [inline node/node_auth_test.go:15]:
```suggestion
```

File: node/node_auth_test.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
+package node
```
Comment (holiman) [inline node/node_auth_test.go:1]: license

File: rpc/http.go [hunk 4]
```diff
@@ -186,6 +211,12 @@ func (hc *httpConn) doRequest(ctx context.Context, msg interface{}) (io.ReadClos
 	hc.mu.Lock()
 	req.Header = hc.headers.Clone()
 	hc.mu.Unlock()
+	if hc.auth != nil {
+		err = hc.auth.AddAuthHeader(&req.Header)
+		if err != nil {
+			return nil, err
+		}
```
Comment (holiman) [inline rpc/http.go:218]:
```suggestion
		if err := hc.auth.AddAuthHeader(&req.Header); err != nil {
			return nil, err
		}
```

File: rpc/websocket.go [hunk 5]
```diff
@@ -224,6 +253,13 @@ func wsClientHeaders(endpoint, origin string) (string, http.Header, error) {
 	if origin != "" {
 		header.Add("origin", origin)
 	}
+	// Add JWT Authorization Header if provided
+	if authProvider != nil {
+		err = authProvider.AddAuthHeader(&header)
+		if err != nil {
```
Comment (holiman) [inline rpc/websocket.go:259]:
```suggestion
		if err := authProvider.AddAuthHeader(&header); err != nil {
```

**Review by holiman — COMMENTED** [review]
Looks pretty good to me, but I'll leave it for @fjl to do a more thorough review

**Comment by trianglesphere** [comment]
One thing that I'd call out is the  ability to potentially add the id/clv claims (these are optional in the spec and TBH I don't know if they are used, but they are there). The id strings would need to be passed into the NewAuth fn.

It's also probably fine to skip this for now until those fields are more commonly used.

File: rpc/client.go [hunk 6]
```diff
@@ -186,6 +186,25 @@ func DialContext(ctx context.Context, rawurl string) (*Client, error) {
 	}
 }
 
+func DialWithAuth(ctx context.Context, rawurl string, auth HeaderAuthProvider) (*Client, error) {
+	u, err := url.Parse(rawurl)
+	if err != nil {
+		return nil, err
+	}
+	switch u.Scheme {
+	case "http", "https":
```
Comment (trianglesphere) [inline rpc/client.go:195]: Returning an error here or in the HTTPProviderWithAuth if the authorprovider is nil might be nice. You'll get an error when doing the connection, but erring here makes it clear what the problem is.

(Also needs to be mirrored for WS)

File: rpc/auth.go [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+type JWTAuthProvider struct {
+	secret []byte
+}
+
+// NewJWTAuthProvider creates a new JWT Auth Provider.
+// The secret should not be empty.
+func NewJWTAuthProvider(jwtsecret []byte) *JWTAuthProvider {
+	return &JWTAuthProvider{secret: jwtsecret}
```
Comment (trianglesphere) [inline rpc/auth.go:27]: This could check some info about the validity of the secret, i.e length - I'm not sure how the error would surface, but surfacing it here might be easier.

File: rpc/client.go (hunk 6 above)
Comment (protolambda) [reply rpc/client.go:195]: `DialHTTPWithAuth` and `DialWebsocketWithAuth` now check if the `auth` is not nil, and I added test cases for it.

File: rpc/auth.go (hunk 7 above)
Comment (protolambda) [reply rpc/auth.go:43]: Changed it to take a `[32]byte` secret instead, avoiding the length edge cases without introducing more errors to handle.

The spec says: "which designates a file containing the hex-encoded 256 bit secret key to be used for verifying/generating jwt tokens." about the cli option, but then also "If such a parameter is given, but the file cannot be read, or does not contain a hex-encoded key of at least 256 bits, the client should treat this as an error".

Internally the jwt library just hashes the key if it's longer than 256 bits, and it seems reasonable to enforce that length based on spec wording.

**Comment by protolambda** [comment]
I think I addressed all open review, let me know if there are other things to test or change.

**Comment by protolambda** [comment]
Note: the time related tests are flaky on a slower machine, since they rely on the existing auth code, which uses `time.Now`. Maybe we can refactor the existing auth code to use a clock that can be mocked, so the test does not rely on time anymore?

**Comment by holiman** [comment]
> Note: the time related tests are flaky on a slower machine, since they rely on the existing auth code, which uses `time.Now`. Maybe we can refactor the existing auth code to use a clock that can be mocked, so the test does not rely on time anymore?

I recently fixed a similar issue here: https://github.com/ethereum/go-ethereum/pull/25120 -- by changing it so that the token is lazily created just before the request is made. That approach is probably simpler than mocking a clock throughout the pipeline?

**Comment by protolambda** [comment]
The auth header creation is already deferred to just before the `client.Do`, the timestamp creation is accurate.

The problem with CI turned out to be different: when testing boundaries with millisecond accuracy you need timestamps to be accurate too. But second-precision was being used, even though JWT supports "NumericDate" format for string timestamps that have higher precision. See https://datatracker.ietf.org/doc/html/rfc7519#section-2 and the `jwt.NumericDate` type.

The precision is now configurable in the client (the server already accepted arbitrary precision), so we can test the boundary accurately. But it's up to the caller to set their own precision here. (not all servers may implement the spec properly).

Also, I fixed an issue where the same auth-headers were being reused for websocket reconnections. Since the auth is time sensitive the auth header needs to be reset.

@holiman this PR is ready for review again.

**Comment by fjl** [comment]
Hey @protolambda, this is just to let you know I haven't forgotten about this PR. Will get to it eventually!

**Comment by protolambda** [comment]
Tests passed, but then got interrupted after I enthusiastically changed pushed a second commit to use `NumericDate` with truncated time based on desired precision, instead of changing the global precision var in the jwt package.

In `NumericDate.UnmarshalJSON` it parses it as float, and converts it into a time in `newNumericDateFromSeconds`, but then constructs the numeric date with `NewNumericDate` instead of `&NumericDate{Time: ...}`. So the server truncates the precision of incoming issued-at claim times, even if the request has a detailed time.

And during `NumericDate.MarshalJSON` it also applies the truncate, so initializing the `NumericDate` doesn't work.

There are a bunch of options:
1. Set the global `jwt.TimePrecision` to a nanosecond, and then the time precision set by the user will always stay untouched on both server/client usage.
2. Set the global `jwt.TimePrecision` to a millisecond, and don't offer the option to the user to change it. If the issued-at bound is 5 seconds, I think it's reasonable to allow millisecond precision, or the 1/5 of the range is enforced by chance of at what time it rounds down.
3. Keep the second time precision, but only override the `jwt.TimePrecision` global in the test, so the test of the 5 second boundary is accurate, even though it's not for the user.

Personally I think option 2 is best; almost no code, no surprise to the user about time precision affecting the request validity, and simple (no options you've to be aware of). Will update the code with that now.

**Comment by protolambda** [comment]
While this PR was stuck the execution API specs changed the expiry time from 5 seconds to 60 seconds. So I'm going to update the tests + revert the seconds->milliseconds precision change.

**Comment by fjl** [comment]
This is ready now. I have change this to use an 'options' scheme after all, because the added client constructor functions were just a bit too much duplication.

Another change: the client-side JWT authentication provider is now also in package `node`. I don't want to put it in package `rpc` because it would add the JWT library dependency to a lot of other packages that don't need it. Also, the tests for this provider were already in package `node`.
